pontoon/checks/libraries/compare_locales.py

```diff
@@ -0,0 +1,184 @@
+    'ComparePropertiesEntity',
+    (
+        'key',
+        'val',
+        'raw_val',
+        'pre_comment',
+    )
+)
+CompareDTDEntity = namedtuple(
+    'CompareDTDEntity',
+    (
+        'key',
+        'val',
+        'raw_val',
+        'pre_comment',
+        'all',
+    )
+)
+
+
+DTD_ENTITY_TMPL = '<!ENTITY %s \"%s\">'
+
+
+class UnsupportedResourceTypeError(Exception):
+    """Raise if compare-locales doesn't support given resource-type."""
+    pass
+
+
+PROPERTIES_ENTITY_ESCAPE_RE = re.compile(
+    r'\\((?P<uni>u[0-9a-fA-F]{1,4})|'
+    r'(?P<nl>\n\s*)|(?P<single>.))',
+    re.M
+)
+PROPERTIES_KNOWN_ESCAPES = {'n': '\n', 'r': '\r', 't': '\t', '\\': '\\'}
+
+
+def unescape_properties_entity(raw_value):
+    """
+    Unescape a raw string.
+    """
+    def unescape(match):
+        found = match.groupdict()
+        if found['uni']:
+            return unichr(int(found['uni'][1:], 16))
+        if found['nl']:
+            return ''
+        return PROPERTIES_KNOWN_ESCAPES.get(found['single'], found['single'])
+    return PROPERTIES_ENTITY_ESCAPE_RE.sub(unescape, raw_value)
+
+
+def cast_to_compare_locales(resource_ext, entity, string):
+    """
+    Cast a Pontoon's translation object into Entities supported by `compare-locales`.
+
+    :arg basestring resource_ext: extension of a resource.
+    :arg pontoon.base.models.Entity entity: Source entity
+    :arg basestring string: a translation
+    :arg pontoon.base.models.Locale locale: Locale of a translation
+    :return: source entity and translation entity that will be compatible a compare-locales checker.
+        Type of those entities depends on the resource_ext.
+    """
+    if resource_ext == '.properties':
+        return (
+            ComparePropertiesEntity(
+                entity.key,
+                unescape_properties_entity(entity.string),
+                entity.string,
+                CommentEntity(entity.comment)
+            ),
+            ComparePropertiesEntity(
+                entity.key,
+                unescape_properties_entity(string),
+                string,
+                CommentEntity(entity.comment),
+            )
+        )
+
+    elif resource_ext == '.dtd':
+        return (
+            CompareDTDEntity(
+                entity.key,
+                html_unescape(entity.string),
+                entity.string,
+                CommentEntity(entity.comment),
+                DTD_ENTITY_TMPL % (entity.key, entity.string)
+            ),
+            CompareDTDEntity(
+                entity.key,
+                html_unescape(string),
+                string,
+                CommentEntity(entity.comment),
+                DTD_ENTITY_TMPL % (entity.key, entity.string)
+            )
+        )
+
+    elif resource_ext == '.ftl':
+        parser = FluentParser()
+
+        parser.readContents(entity.string.encode('utf-8'))
+        refEntity, = list(parser)
+
+        parser.readContents(string.encode('utf-8'))
```

Comment:
@Pike @jotes Do we [still](https://bugzilla.mozilla.org/show_bug.cgi?id=1448256) need to `encode` here and above now that we use compare-locales 3.1?

Reply:
Good catch, this can use `readUnicode` now. There's a final IO and encoding related fix in 3.1.1, calling that out, though it probably won't affect pontoon.